code/game/objects/structures/doors.dm

```diff
@@ -0,0 +1,902 @@
+	return TRUE
+
+/obj/structure/door/multitool_act(mob/living/user, obj/item/I)
+	return handle_tools(user, I)
+
+/obj/structure/door/analyzer_act(mob/living/user, obj/item/I)
+	return handle_tools(user, I)
+
+///Reduce line copy and pasting by having one proc do (most of) them
+/obj/structure/door/proc/handle_tools(mob/living/user, obj/item/tool)
+	if(!length(construction_steps))	//No message needed because the user would know if they examined the door!
+		return FALSE
+
+	if(!CHECK_BITFIELD(door_flags, DOOR_FORCEABLE))
+		to_chat(user, span_warning("[src] is too sturdy to disassemble!"))
+		return FALSE
+
+	//Don't allow de/construction if the door is closed and locked
+	if(lock?.locked && CHECK_BITFIELD(door_flags, DOOR_OPEN))
+		balloon_alert(user, "Unlock it first!")
+		return FALSE
+
+	//Check if the tool can deconstruct the door
+	if(tool.tool_behaviour != construction_steps[construction_steps_index])
+		//Check if the tool can construct the door; make sure list[0] isn't done by accident
+		if(construction_steps_index == 1 || (tool.tool_behaviour != construction_steps[construction_steps_index - 1]))
+			return FALSE
+
+		if(tool_action_time && !do_after(user, tool_action_time, NONE, src, BUSY_ICON_BUILD))
+			return FALSE
+
+		//Checks passed, construct one step
+		do_construction()
+		return TRUE
+
+	if(tool_action_time && !do_after(user, tool_action_time, NONE, src, BUSY_ICON_BUILD))
+		return FALSE
+
+	//Checks passed, deconstructed one step
+	do_deconstruction()
+	return TRUE
+
+///Deconstruct the door
+/obj/structure/door/proc/do_deconstruction()
+	var/number_of_steps = length(construction_steps)
+	construction_steps_index++
+
+	if(construction_steps_index == number_of_steps)
+		name = "[base_name] frame"
+		anchored = FALSE
+		return	//Don't need to keep going if this condition was met
+
+	if(construction_steps_index > number_of_steps)
+		deconstruct(TRUE)
+
+///Build the door
+/obj/structure/door/proc/do_construction()
+	var/number_of_steps = length(construction_steps)
+	construction_steps_index--
+
+	if(construction_steps_index != number_of_steps - 1)
+		name = base_name
+		anchored = initial(anchored)
+
+/obj/structure/door/proc/toggle_lock(mob/user, obj/item/key/door/key)
+	if(!lock)
+		balloon_alert(user, "No lock installed")
+		return
+
+	lock.toggle(user, key)
+
+/obj/structure/door/Cross(atom/movable/AM)
+	. = ..()
+	if(.)	//Non-full tile doors will not block the crosser, no need to continue below
+		if(CHECK_BITFIELD(atom_flags, ON_BORDER) && REVERSE_DIR(dir) == AM.dir)	//Auto close non-full tile doors if the crosser walked through the door
+			auto_close(AM)
+		return .
+
+	//Don't care about the user if they're not a mob type
+	attempt_to_open(ismob(AM) ? AM : null, direction_from_opener = angle2dir(Get_Angle(src, AM)), bumped = TRUE)
+	var/is_open = CHECK_BITFIELD(door_flags, DOOR_OPEN)
+	if(is_open && CHECK_BITFIELD(atom_flags, ON_BORDER) && REVERSE_DIR(dir) == AM.dir)
+		auto_close(AM)
+
+	return is_open
```

Comment:
isnt this true when the door autocloses? is that intended?

Reply:
Nope, the intent of `return is_open` is to allow for people tackling through doors to keep walking. If `attempt_to_open` instantly opened the door, `is_open` will be true. Otherwise, the door is in the process of opening but not open yet. The `auto_close` bit is just to close a non-full tile door (think windoors) if you walk into it but was already open so that it starts closing behind you. I guess I should have commented this bit. 😅